fix Super high COP in VRFFluidTCtrl model

yujiex opened the pull request:
Pull request overview
---------------------
 - Fixes #10751
 - Fixes #10764
 - Fixes #10767
 
### 1. Fix of the super high COP problem

 The following shows the distribution of heating and cooling COP before vs after the fix (figure updated Mar. 11). For display purpose, the COP=0 points are removed.
![image](https://github.com/user-attachments/assets/c2b0d99f-bb93-4e78-9e19-bc63b4c10632)
[eplusout_defect_output_1025.csv](https://github.com/user-attachments/files/17525725/eplusout_defect_output_1025.csv)
[eplusout_fix_0311.csv](https://github.com/user-attachments/files/19196143/eplusout_fix_0311.csv)

COP is calculated as `this->TotalHeatingCapacity` divided by `this->ElecHeatingPower` and other auxiliary power. 
```
this->OperatingHeatingCOP = (this->TotalHeatingCapacity) /
                            (this->ElecHeatingPower + this->CrankCaseHeaterPower + this->EvapCondPumpElecPower + this->DefrostPower);
```
`this->ElecHeatingPower` is calculated as compressor power and outdoor fan power `this->ElecHeatingPower = this->Ncomp + this->OUFanPower;`

The following shows how `this->TotalHeatingCapacity` and `this->TotalCoolingCapacity`  are calculated in FluidTCtrl model and the curve-based model. This has led me to believe the super high COP is cause by numerator not multiplying the `CyclingRatio` term. **But it's NOT because of this.** 
```
// FluidTCtrl model
    this->TotalCoolingCapacity = TotalCondCoolingCapacity * CoolingPLR;
    this->TotalHeatingCapacity = TotalCondHeatingCapacity * HeatingPLR;
// curve-based model
    vrf.TotalCoolingCapacity = TotalCondCoolingCapacity * CoolingPLR * CyclingRatio;
    vrf.TotalHeatingCapacity = TotalCondHeatingCapacity * HeatingPLR * CyclingRatio;
```
#### heating COP fix
 
Even if here the cycling ratio is not multiplied, when plotting the "VRF Heat Pump Total Heating(Cooling) Rate" against cycling ratio, there's actually a positive linear relationship. However, the linear relationship between heating-rate and cycling ratio has a noticeable intercept, while the intercept for the compressor power and cycling ratio line is near zero. This would mean, when cycling ratio is very small, numerator in heating COP will be non-zero while the denominator approaches zero (compressor power and fan power both go to zero), the heating COP will thus blow up.

<img width="884" alt="image" src="https://github.com/user-attachments/assets/7baa78f1-5921-4ce2-af19-fc90d1d8a272">

Comparing with v23.1, this non-zero intercept is also present in the heat pump heating rate and cycling ratio relationship. 
<img width="867" alt="image" src="https://github.com/user-attachments/assets/dcd04e38-bfca-4b77-a8c9-c407ccdd9c37">

However, it didn't blow up as the outdoor unit fan power is always constant and not cycling with the coil even if the "Supply Air Fan Operating Mode Schedule Name" points to a constant-zero schedule (fan always cycles with coil). The following shows the comparison of heatingCOP vs cycling ratio relationship between current develop and v23-1. We can see when zoomed in, they both have the exponential decay pattern.

![image](https://github.com/user-attachments/assets/38eaa17d-6e21-4805-b347-70fb117242e1)

To fix it, the Ncomp multiplying by cycling ratio is moved to inside of the `VRFOU_CalcCompH` function. Previously (in the develop branch), Ncomp is calculated inside the function by multiplying the rated capacity and the power performance curve value, and the cycling ratio is multiplied to the calculated Ncomp after the `VRFOU_CalcCompH`. However, since the heating demand in `VRFOU_CalcCompH` involves the compressor heat release (`Q_evap_req = TU_load + Pipe_Q - Ncomp;`), if here `Ncomp` didn't multiply by cycling ratio, than the compressor heat release is overestimated, leading to an underestimation of the heating demand `Q_evap_req`, as a result, the computed cycling ratio will also be smaller than it should be, creating a mismatch between the cycling ratio and the heating rate. Whether the cycling ratio is multiplied inside or after `VRFOU_CalcCompH` will matter in the cycling ratio calculation in heating mode, but not in cooling mode as Ncomp was not in the demand calculation there.

After applied this fix, this is how the heating rate, compressor power, and COP look relative to cycling ratio

<img width="860" alt="image" src="https://github.com/user-attachments/assets/7e3d364e-6009-4a24-8d7f-d5a7b9b6794b">

#### cooling COP fix

The cooling COP didn't go up as high as heating one but still with a maximum of 150+, which is too high to be realistic. The reason for the high cooling COP is likely the non-linearity between heat pump cooling rate and the cycling ratio when cycling ratio is very small. 

<img width="877" alt="image" src="https://github.com/user-attachments/assets/60c13a95-0d75-485e-9736-0e912735aca6">

In the cycling ratio calculation, the code attempts to match the demand and capacity. If the they cannot match because the capacity is too large, then it will cycle and the cycling ratio is calculated as 

```
Cap_Eva0 = (TU_load + Pipe_Q) * C_cap_operation; // <- demand
Cap_Eva1 = this->CoffEvapCap * this->RatedEvapCapacity *
            CurveValue(state, this->OUCoolingCAPFT(CounterCompSpdTemp), T_discharge, T_suction); // <- capacity
CyclingRatio = Cap_Eva0 / Cap_Eva1;
```

When the cycling ratio is very small, this adjustment term `C_cap_operation` exhibits a exponential decay pattern as well and quickly goes to 0 as cycling ratio drops, which seems to be the cause of the non-linear pattern in the heat pump cooling rate-cycling ratio relationship (the following plot is in the develop branch, using debug prints to export value of cyclingRatio and C_cap_operation after line 11486 (`Q_h_OU *= CyclingRatio;` line)). The 6 subfigures on the right show how the 6 input variables vary with cycling ratio. `h_comp_in_real`, `P_evap_reql`, and `T_comp_in_rate` show some shoot up very high for when the cycling ratio is low.

![image](https://github.com/user-attachments/assets/15aaa1e9-c2e3-41c4-806d-7959c42b2d32)

Note that this multiplier `C_cap_operation` only trends to zero in cooling mode, not in heating mode. Maybe a threshold should be placed on this in cooling mode.

![cyclingRatio_vs_C_cap_operation_heatcool_dev](https://github.com/user-attachments/assets/6ba07694-fd4c-4ca2-a99b-cb35553fdc3f)

after removing this multiplier, the cooling rate non-linear behavior is gone and COP vs cycling ratio plot is like the following

<img width="880" alt="image" src="https://github.com/user-attachments/assets/f2f887c6-39ae-4320-8181-245df711ca7f">

`C_cap_operation` is calculated in function `VRFOU_CapModFactor`, as `C_cap_operation = C_cap_density * C_cap_enthalpy;` in the test case (using ), `C_cap_density` is steadily around 1. `C_cap_enthalpy` exponential decays from 1 to 0 at low cycling ratio. So the problem is likely in `C_cap_enthalpy`.

![image](https://github.com/user-attachments/assets/6726f25b-9c00-4ccb-bd7c-175e6e9a6f00)

`C_cap_enthalpy` is calculated as `C_cap_enthalpy = std::abs(h_evap_out_rate - h_evap_in_rate) / std::abs(h_comp_in_real - h_evap_in_real);`. Numerator roughly ranges from 180000 to 230000 linearly changing with cycling ratio. The denominator has the exponential decay pattern. `h_evap_in_real` (or `Pipe_h_IU_in`) linearly changes while `h_comp_in_real` or `Pipe_h_comp_in` has the exponential decay pattern. 

`Pipe_h_comp_in` is calculated as `Pipe_h_comp_in = Pipe_h_IU_out + Pipe_Q / Pipe_m_ref;` in the following function

```
                    // Re-calculate piping loss
                    this->VRFOU_PipeLossC(state,
                                          Pipe_m_ref,
                                          max(min(Pipe_Pe_assumed, RefPHigh), RefPLow),
                                          Pipe_h_IU_out,
                                          Pipe_SH_merged,
                                          OutdoorDryBulb,
                                          Pipe_Q,
                                          Pipe_DeltP,
                                          Pipe_h_comp_in); <- calculated here
```

The super high `Pipe_h_comp_in` is related to low refrigerant rate. When the refrigerant flow is very small, the ratio of `Pipe_Q / Pipe_m_ref` shoots up very high, making `Pipe_h_comp_in` very large. From this investigation, it seems the calculation of piping loss might not be correct when refrigerant flow rate is very small.

![pipingLoss_refrig_flow](https://github.com/user-attachments/assets/2d35da44-c4c5-4c4c-a217-a19660991b96)
![pipingLoss_refrig_flow_small_flow](https://github.com/user-attachments/assets/bf32ac78-2e78-4977-ae26-a6c00ddef56e)

Essentially `Pipe_h_comp_in = Pipe_h_IU_out + Pipe_Q / Pipe_m_ref` assumes that the heat loss through the pipe is due to the flow of the refrigerant and that 

heat loss through the pipe = enthalpy difference * refrigerant flow rate

But this is problematic when refrigerant flow rate is very small and close to zero. For example, in the following scatter plot, roughly there's a positive linear relationship between refrigerant flow rate (horizontal axis) and the pipe heat loss (Pipe_Q on the vertical axis). When refrigerant flow is near zero, Pipe_Q is not zero. `Pipe_Q / Pipe_m_ref` this term will blow up and create super high `Pipe_h_comp_in`.

![pipingLoss_refrig_flow](https://github.com/user-attachments/assets/4e1baa36-6a0b-40f1-b39a-690a18f583a2)

The reason for low refrigerant flow rate is due to the small load: state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalCoolLoad(NumTU)
```
                Pipe_m_ref_i = (state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalCoolLoad(NumTU) <= 0.0)
                                   ? 0.0
                                   : (state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalCoolLoad(NumTU) / (Pipe_h_IU_out_i - Pipe_h_IU_in));
                Pipe_m_ref = Pipe_m_ref + Pipe_m_ref_i;
```

According to Daikin engineers, once compressor reach its min speed, refrigerant mass flow will not drop any further. So the following bound is added in the refrigerant flow calculation

```
Real64 min_speed_capacity =
    this->CoffEvapCap * this->RatedEvapCapacity * CurveValue(state, this->OUCoolingCAPFT(1), T_discharge, T_suction);
Pipe_m_ref_i = (state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalCoolLoad(NumTU) <= 0.0)
                    ? 0.0
                    // refrigerant flow rate do not get smaller when the compressor needs to cycle.
                    : max(min_speed_capacity,
                          (state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalCoolLoad(NumTU)) /
                              (Pipe_h_IU_out_i - Pipe_h_IU_in));
```

----

The fact that cooling COP is a bit high might be how the performance data is set up.

The following shows the range of COP using the current range of condensing and evaporating temperature

```
heating mode
+        x            y           cyclingRatio        fanpower   comp.power.calc cooling.rate.calc      COP       
 Min.   :35   Min.   :-15.00   Min.   :0.01308   Min.   :119   Min.   :167.3   Min.   : 797.1    Min.   :2.455  
 1st Qu.:36   1st Qu.: -7.25   1st Qu.:0.27588   1st Qu.:119   1st Qu.:202.1   1st Qu.:1099.9    1st Qu.:3.277  
 Median :37   Median :  0.50   Median :0.49665   Median :119   Median :209.8   Median :1522.8    Median :4.500  
 Mean   :37   Mean   :  0.50   Mean   :0.50749   Mean   :119   Mean   :209.3   Mean   :1612.5    Mean   :4.957  
 3rd Qu.:38   3rd Qu.:  8.25   3rd Qu.:0.72511   3rd Qu.:119   3rd Qu.:219.0   3rd Qu.:2077.6    3rd Qu.:6.373  
 Max.   :39   Max.   : 16.00   Max.   :0.99268   Max.   :119   Max.   :230.5   Max.   :2797.2    Max.   :9.769  

cooling mode
+        x            y        cyclingRatio        fanpower   comp.power.calc cooling.rate.calc      COP        
 Min.   :16   Min.   :6.0   Min.   :0.01339   Min.   :119   Min.   : 71.7   Min.   :1920      Min.   : 5.899  
 1st Qu.:21   1st Qu.:6.0   1st Qu.:0.28550   1st Qu.:119   1st Qu.:106.2   1st Qu.:2080      1st Qu.: 7.109  
 Median :26   Median :6.5   Median :0.53528   Median :119   Median :140.2   Median :2211      Median : 8.534  
 Mean   :26   Mean   :6.5   Mean   :0.53092   Mean   :119   Mean   :140.0   Mean   :2209      Mean   : 8.833  
 3rd Qu.:31   3rd Qu.:7.0   3rd Qu.:0.77554   3rd Qu.:119   3rd Qu.:173.6   3rd Qu.:2335      3rd Qu.:10.370  
 Max.   :36   Max.   :7.0   Max.   :0.99191   Max.   :119   Max.   :206.5   Max.   :2495      Max.   :13.085  

```
The data files for the above summary is here
[dummuy_heating_data.csv](https://github.com/user-attachments/files/18592943/dummuy_heating_data.csv)
[dummuy_cooling_data.csv](https://github.com/user-attachments/files/18592944/dummuy_cooling_data.csv)


Note that in version 23.1, the cycling ratio is like the following

![image](https://github.com/user-attachments/assets/bfa24f15-36ed-48ab-8e3b-d038855a8425) 

### 2. Fix of the refrigerant warning

The issue is originated from the change of this line in PR#10416 the following line. The intention is to 

![image](https://github.com/user-attachments/assets/658f3593-a4d0-4410-9c4c-95a7b0c33a89)

The intention of the change is to impose a lower bound in solving for an evaporating temperature when the heating load is small. Such a lower bound is necessary as otherwise, in many time steps when the system is cycling, the OU evaporating temperature will be -72C, which is too low to be feasible. However, using the value of the input field "Variable Evaporating Temperature Minimum for Indoor Unit" as the lower bound will trigger a bunch of refrigerant warnings.

To fix the refrigerant flow problem, it is observed that the following line might have been a mistake `CapMinPe = min(Pdischarge - this->CompMaxDeltaP, RefMinPe);`. `Pdischarge` is the compressor discharge pressure,
`this->CompMaxDeltaP` is the maximum compressor pressure rise. Then `Pdischarge - this->CompMaxDeltaP` is minimum compressor suction pressure considering pressure rise bound. The `RefMinPe` is minimum refrigerant evaporating pressure, which is a pressure lower bound considering physical feasibility. `CapMinPe` should be no less than both of them, so it should be max of them, not min.

In the code base, there are three places where `CapMinPe` is calculated (shown in the following). Two of them are max, one is min, hinting that the min might be a bug.

![image](https://github.com/user-attachments/assets/57ff6e88-80ce-4bd0-bd52-3baa54fb5aa6)

After imposing the right pressure bound, the refrigerant warnings are mostly gone. The following are the eplusout.err files from the develop and the feature branch. A difference can be seen just from the file size.

[fix_eplusout.err.zip](https://github.com/user-attachments/files/17122178/fix_eplusout.err.zip)
[develop_eplusout.err.zip](https://github.com/user-attachments/files/17122180/develop_eplusout.err.zip)

### 3. Fix of the recurring warning issue

The warning message keep showing up is the non-recurring warning (which should only be shown once or a few times).

The excessive amount of warning is produced by this chunk of code. The condition `this->errors[(int)RefrigError::SatSupDensity].count <= df->RefrigErrorLimitTest` decides whether it prints this warning. The right hand side of the inequality is 1, while the left hand side overflows to negative at some point to -2147450880. From then on, the condition just keeps evaluates to true and the warning keeps showing up. It overflows as it is not a counter, but a sum of all previous counter values. This makes it get large very quickly. When it overflows, the counter value of `df->SatErrCountGetSupHeatDensityRefrig` is 65536. This `this->errors[(int)RefrigError::SatSupDensity].count <= df->RefrigErrorLimitTest` is 1 + 2 + ... + 65536, which overflowed.

```
++df->SatErrCountGetSupHeatDensityRefrig;
// send warning
this->errors[(int)RefrigError::SatSupDensity].count += df->SatErrCountGetSupHeatDensityRefrig;
// send warning
if (this->errors[(int)RefrigError::SatSupDensity].count <= df->RefrigErrorLimitTest) {
    ShowWarningMessage(
        state,
        format("{}: Refrigerant [{}] is saturated at the given conditions, saturated density at given temperature returned. **",
                routineName,
                this->Name));
    ShowContinueError(state, fmt::format("...Called From:{}", CalledFrom));
    ShowContinueError(state, format("Refrigerant temperature = {:.2R}", Temperature));
    ShowContinueError(state, format("Refrigerant pressure = {:.0R}", Pressure));
    ShowContinueError(state, format("Returned Density value = {:.3R}", saturated_density));
    ShowContinueErrorTimeStamp(state, "");
}
```

### 4. Cooling Coil Type [1] IP to SI unit conversion warning

The warning is produced because the code assumes there is a unit of measure inside the bracket and will attempt to perform unit conversion. Here this is a categorical variable and should not attempt to do unit conversion, so [] is not appropriate. The problem if fixed by changing the "[]" into "()" in this "Cooling Coil Type [1]".

### Regression diffs

#### Table big diff and Table string diffs

There are 733 files have "ColumnHeadingDifference", leading to table big diffs and string diffs. This is caused by changing "Cooling Coil Type [1]" to "Cooling Coil Type (1)" in this feature branch. The diffs are expected.

#### meter diff

Meter diff happens in test idf "VariableRefrigerantFlow_FluidTCtrl_5Zone", in the following two fields.
- Electricity:Facility
- Electricity:HVAC    
The meter diff is caused by this change. The changes here will cause differences in compressor power, which will lead to changes in Electricity:HVAC and Electricity:Facility.
![image](https://github.com/user-attachments/assets/067edbeb-5319-447c-ab6d-507dd9390a9f)
 
#### eso diff

in US+SF+CZ4A+hp+crawlspace+IECC_2006_VRF, VariableRefrigerantFlow_FluidTCtrl_5Zone, and VariableRefrigerantFlow_FluidTCtrl_HR_5Zone, the following fields have eso diffs

- VRF HEAT PUMP:VRF Heat Pump Total Cooling Rate
- VRF HEAT PUMP:VRF Heat Pump Total Heating Rate
- VRF HEAT PUMP:VRF Heat Pump Heating Electricity Rate
- VRF HEAT PUMP:VRF Heat Pump Heating Electricity Energy
- VRF HEAT PUMP:VRF Heat Pump Compressor Electricity Rate
- VRF HEAT PUMP:VRF Heat Pump Outdoor Unit Evaporating Temperature

These diffs are expected as the numerator of the COP term (heating/cooling output) is changed in this feature: it is multiplied by the cycling ratio now. The changes in COP and heating cooling rate are both expected. Changes in Outdoor Unit Evaporating Temperature is due to the correction in refrigerant pressure lower bound, which is used in calculation of OU evaporating temperature. Heating electricity rate and energy difference happen in 1/21. Heating electricity rate is the sum of compressor power and outdoor unit fan power. As a result of the changing `MinOutdoorUnitTe` argument from `max(this->IUEvapTempLow, CapMinTe)` to `CapMinTe` in function `VRFOU_CalcCompH`, compressor power will change. So diffs in heating electricity rate is also expected.

#### err diffs

11 files have error diffs. This is because the following warning is removed as a result of the column header change of "Cooling Coil Type".

```
** Warning ** Unable to find a unit conversion from Cooling Coil Type [1] into IP units
```

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Commit:
fix TotalHeatingCapacity, it missed cycling ratio

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -12198,8 +12198,8 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state, c
-    this->TotalCoolingCapacity = TotalCondCoolingCapacity * CoolingPLR;
-    this->TotalHeatingCapacity = TotalCondHeatingCapacity * HeatingPLR;
+    this->TotalCoolingCapacity = TotalCondCoolingCapacity * CoolingPLR * CyclingRatio;
+    this->TotalHeatingCapacity = TotalCondHeatingCapacity * HeatingPLR * CyclingRatio;
```
OK, I can see this one being missed because it's multiplied by PLR and looked correct at first glance. If COP came back in line then it's likely that all the electricity reports are correctly adjusting for cycling. Although it's not a given, because things like defrost, cranckcase heater and evap condenser pump power are small compared to compressor power, and are most likely 0 in the test files. I would check those.  What else might have been missed? What happens to piping losses during cycling? What about something like compressor speed? should that also be adjusted by CyclingRatio?

yujiex replied:
Crankcase heater seems to have considered cycling
```
        VRFRTF = min(1.0, (CyclingRatio / PartLoadFraction));
...
        this->CrankCaseHeaterPower = this->CCHeaterPower * (1.0 - VRFRTF);
```
Defrost seems to have its own "runtime fraction", the `FractionalDefrostTime`: 
```
                    this->DefrostPower = DefrostEIRTempModFac * (this->HeatingCapacity / 1.01667) * FractionalDefrostTime;

                } else { // Defrost strategy is resistive
                    this->DefrostPower = this->DefrostCapacity * FractionalDefrostTime;
                }
```

rraustad replied:
Crankcase heater power looks good. I think defrost power needs a  CyclingRatio term. FractionalDefrostTime is the defrost time for a full time step. If the system doesn't run that time step (i.e., CyclingRatio = 0.000000001) then DefrostPower should be 0. You can check FractionalDefrostTime and see that it's a function of OA humidity ratio and has nothing to do with PLR or CyclingRatio. Of course you have to check my thinking before doing anything.

yujiex replied:
I think you're right. FractionalDefrostTime is either from the user input or from `OutdoorCoildw`. It should probably multiply by cycling ratio.

For `this->EvapCondPumpElecPower` term, since the FluidTCtrl model doesn't have the "Evaporative Condenser Pump Rated Power Consumption" field like the curve-based VRF, this term is always 0.

Commit:
Consider cycling in defrost power

Commit:
Merge remote-tracking branch 'origin/develop' into fixVRFhighCOP

Commit:
Here VRFRTF is not initialized, don't multiply here to defrost

Commit:
move defrost RTF adjust out of the if block

so that when the outdoor conditions cause a non-zero defrost power, AND the
system if off, defrost power would be set to zero.

Commit:
move defrostPower adjustment after VRFRTF is initialized

Commit:
Fix min pressure lower bound

Pdischarge is the compressor discharge pressure,
this->CompMaxDeltaP is the maximum compressor pressure rise
Pdischarge - this->CompMaxDeltaP is minimum compressor suction pressure

RefMinPe is minimum refrigerant evaporating pressure

These two terms are both lower bounds of the suction pressure

CapMinPe should be no less than both of them, so CapMinPe should be max of the
two LB, not min

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -1192,10 +1192,9 @@ void CalcVRFCondenser(EnergyPlusData &state, int const VRFCond)
-        vrf.DefrostPower *= VRFRTF;
     }
+    // adjust defrost power based on RTF
+    vrf.DefrostPower *= VRFRTF;
```
This looks correct.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -11592,7 +11591,8 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state, c
         // Evaporative capacity ranges_Min
-        CapMinPe = min(Pdischarge - this->CompMaxDeltaP, RefMinPe);
+        // suction pressure lower bound need to be no less than both terms in the following
+        CapMinPe = max(Pdischarge - this->CompMaxDeltaP, RefMinPe);
```
This looks correct. Not sure why this didn't show up during the original model develolpment.

yujiex replied:
Yeah, I'm surprised as well.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -12237,6 +12236,7 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state, c
         this->ElecHeatingPower = 0;
     }
     this->VRFCondRTF = VRFRTF;
+    this->DefrostPower *= VRFRTF;
```
This looks correct. A plot of DefrostPower vs RTF or CyclingRatio would show it's working.

yujiex replied:
This is a plot of heat pump defrost power vs cycling ratio

![image](https://github.com/user-attachments/assets/b3a34a77-e457-4a01-b56d-e34d33b41e17)

When the defrost power is non-zero, it is proportional to cycling ratio

The output file is here:
[eplusout_defrost_vs_cyclingRatio.xlsx](https://github.com/user-attachments/files/17135490/eplusout_defrost_vs_cyclingRatio.xlsx)

rraustad replied:
Why is defrost electricity so small? at 6 E -9

yujiex replied:
I used this test file. Its "Resistive Defrost Heater Capacity" is 0.0000001
 [US+SF+CZ5B+hp+slab+IECC_2021_VRFPhysics_v2_hrdsize_V2420.expidf.zip](https://github.com/user-attachments/files/17122056/US%2BSF%2BCZ5B%2Bhp%2Bslab%2BIECC_2021_VRFPhysics_v2_hrdsize_V2420.expidf.zip)

rraustad commented:
This branch looks good now. I looked at fix_eplusout.err.zip and see this warning. Why is OutputReportTabular looking to convert a cooling coil type to IP units? This is probably a FluidTCtrl cooling coil. You could fix that here or file an issue.

    ** Warning ** Unable to find a unit conversion from Cooling Coil Type [1] into IP units

rraustad commented:
@yujiex these warnings (develop_eplusout.err.zip) are set up as recurring warnings so that the err file does not contain thousands of messages. However, this file does contain thousands of messages. Can you please investigate, either here or a different branch.

    int RefrigErrorLimitTest = 1; // how many times error is printed with details before recurring called

    ++df->SatErrCountGetSupHeatDensityRefrig;
    // send warning
    this->errors[(int)RefrigError::SatSupDensity].count += df->SatErrCountGetSupHeatDensityRefrig;
    // send warning
    if (this->errors[(int)RefrigError::SatSupDensity].count <= df->RefrigErrorLimitTest) {
        ShowWarningMessage(

    ** Warning ** RefrigProps::getSupHeatDensity: Refrigerant [R410A] is saturated at the given conditions, saturated density at given temperature returned. **
    **   ~~~   ** ...Called From:VRFOU_CapModFactor
    **   ~~~   ** Refrigerant temperature = -3.55
    **   ~~~   ** Refrigerant pressure = 962140
    **   ~~~   ** Returned Density value = 27.216
    **   ~~~   **  Environment=ANNUAL, at Simulation time=01/03 16:50 - 17:00

yujiex commented:
> Unable to find a unit conversion from Cooling Coil Type [1] into IP units

Thanks for noticing this. I will fix it here too

Commit:
fix warning of Cooling Coil Type SI to IP unit conversion

Commit:
resolve err counter integer overflow

the error message counter is a sum of all previous counter values
it overflows to negative
non-recurring warning keeps showing up after the overflow
use the counter, not the sum of counter in the predicate

yujiex commented:
@rraustad I've investigated the two issues

**SI to IP unit conversion warning**

I fixed the warning of "unit conversion from Cooling Coil Type [1] into IP units", by changing the [] to () in the column header "Cooling Coil Type [1]". The warning is produced because the code assumes there is a unit of measure inside the bracket and will attempt to perform unit conversion. Here this is a categorical variable and should not attempt to do unit conversion, so [] is not appropriate. It is changed to () per suggestion from @JasonGlazer 

**recurring warning shows too many times**

I also figured out the reason for the recurring warning message not showning as recurring. The messages keeps showing up is not produced by the `ShowRecurringWarningErrorAtEnd`, but by the following non-recurring part. The condition `this->errors[(int)RefrigError::SatSupDensity].count <= df->RefrigErrorLimitTest` decides wiether it will produce this warning. The right hand side of the inequality is 1, the left hand side overflows to negative at some point to -2147450880, and from then on, the condition just keeps evaluates to true and the warning keeps showing up. It overflows as it is not a counter, but a sum of all previous counter values. When it overflows, the counter value of `df->SatErrCountGetSupHeatDensityRefrig` is 65536. This `this->errors[(int)RefrigError::SatSupDensity].count <= df->RefrigErrorLimitTest` is 1 + 2 + ... + 65536, which overflowed.

I believe changing the condition to just `if (df->SatErrCountGetSupHeatDensityRefrig <= df->RefrigErrorLimitTest)` will solve the problem
```
++df->SatErrCountGetSupHeatDensityRefrig;
// send warning
this->errors[(int)RefrigError::SatSupDensity].count += df->SatErrCountGetSupHeatDensityRefrig;
// send warning
if (this->errors[(int)RefrigError::SatSupDensity].count <= df->RefrigErrorLimitTest) {
    ShowWarningMessage(
        state,
        format("{}: Refrigerant [{}] is saturated at the given conditions, saturated density at given temperature returned. **",
                routineName,
                this->Name));
    ShowContinueError(state, fmt::format("...Called From:{}", CalledFrom));
    ShowContinueError(state, format("Refrigerant temperature = {:.2R}", Temperature));
    ShowContinueError(state, format("Refrigerant pressure = {:.0R}", Pressure));
    ShowContinueError(state, format("Returned Density value = {:.3R}", saturated_density));
    ShowContinueErrorTimeStamp(state, "");
}
```

Commit:
fix another overflow place, use this->errors[].count as counter

Commit:
clang-format

Commit:
Merge remote-tracking branch 'origin/develop' into fixVRFhighCOP

Commit:
Merge remote-tracking branch 'origin/develop' into fixVRFhighCOP

yujiex commented:
The fix for the high COP part (https://github.com/NREL/EnergyPlus/issues/10751) might have some problem. Let me investigate some more. Please hold off on reviewing. Thanks.

Commit:
Cycling ratio need to be applied inside VRFOU_CalcCompH

because the demand is roughly Q_evap_req = TU_load + Pipe_Q - Ncomp (multiplied
by an adjustment factor C_cap_operation). Ncomp is an input-ouptut variable of
the function. Previously Cycling ratio is multiplied outside of VRFOU_CalcCompH
after the function has computed Ncomp. However, this will lead to very small
cycling ratio non-compatible with the demand as the demand is underestimated
because of the Ncomp (in reality it should not be this large, it should be
Ncomp * CyclingRatio). This doesn't matter on the cooling side as Ncomp is not
involved in the demand calculation there.

Commit:
remove C_cap_operation in demand calculation in cyclingRatio

remove C_cap_operation in demand calculation

Commit:
Merge remote-tracking branch 'origin/develop' into fixVRFhighCOP

Commit:
fix unit test

Commit:
Revert "remove C_cap_operation in demand calculation in cyclingRatio"

This reverts commit 0554057a1ea2bb32d183dc94178d55d73bf99ae3.

nrel-bot-2 commented:
@yujiex it has been 16 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 8 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

Commit:
remove evaporating temperature value passing

Commit:
Revert "remove evaporating temperature value passing"

This reverts commit 5b759a7c382303f06429b5ad0664cc3be37bc7c9.

Myoldmopar commented:
@yujiex let me know what you think about this PR status.  If it's worth pushing it forward, let me know how I can help.

yujiex commented:
> @yujiex let me know what you think about this PR status. If it's worth pushing it forward, let me know how I can help.

Hi @Myoldmopar, the remaining issue to be fixed here is the cooling COP being too high (can reach 158). I have been investigating it and feels the piping loss calculation at low refrigerant flow rate had some issue and I proposed some fix here: [potential fix](#potential-fix-to-cooling-cop). I haven't implemented this fix in the code yet. But conceptually I feel it is likely to work. I'll try it out tomorrow afternoon.

Commit:
add Pipe_h_comp_in calculation at stationary (near zero) flow

Commit:
Merge remote-tracking branch 'origin/develop' into fixVRFhighCOP

Commit:
fix unit tests

yujiex commented:
@rraustad and @mjwitte I have investigated the cooling high COP issue and I think it's related to how the piping loss is calculated at low-flow-rate calculation (explained in section "potential fix to cooling COP" in the PR introduction). Do you think this fix make sense? I have to choose a cutoff value for when it should be considered stationary flow but didn't find much info on the internet about what this cutoff should be. Any ideas on this? Thanks.

Commit:
restrict refrigerant flow to not reduce beyond min compressor speed

Commit:
Merge remote-tracking branch 'origin/develop' into fixVRFhighCOP

yujiex commented:
Hi @mjwitte and @rraustad, I've added a bound for the refrigerant flow rate. As according to Daikin engineers, once compressor reach its min speed, refrigerant mass flow will not drop any further. So the following bound is added in the refrigerant flow calculation.
```
Real64 min_speed_capacity =
    this->CoffEvapCap * this->RatedEvapCapacity * CurveValue(state, this->OUCoolingCAPFT(1), T_discharge, T_suction);
Pipe_m_ref_i = (state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalCoolLoad(NumTU) <= 0.0)
                    ? 0.0
                    // refrigerant flow rate do not get smaller when the compressor needs to cycle.
                    : max(min_speed_capacity,
                          (state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalCoolLoad(NumTU)) /
                              (Pipe_h_IU_out_i - Pipe_h_IU_in));
```
Please see if this make sense. Thanks!

Commit:
fix unit test

Commit:
Revert "fix unit test"

This reverts commit ef8a5e9cba6918ec5fbc52c7d5d7513b1fd99726.

Commit:
fix unit test

Commit:
Merge remote-tracking branch 'origin/develop' into fixVRFhighCOP

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -14016,10 +14015,14 @@ void VRFCondenserEquipment::VRFOU_CalcCompC(EnergyPlusData &state,
+                                                   // refrigerant flow rate do not get smaller when the compressor needs to cycle.
+                                                   : max(min_speed_capacity,
+                                                         (state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalCoolLoad(NumTU)) /
+                                                             (Pipe_h_IU_out_i - Pipe_h_IU_in));
```
This may be the very change that corrects several issues with this model.

yujiex replied:
This fixed the super high cooling COP problem: from previous maximum of 158 to about 12 now.

Myoldmopar reviewed (approved):
I'll buy it.  I admittedly don't have the time to dive into every bit of detail here.  I'm grateful that this has gone through several rounds of reviews and reworks over the last few months, so a big thanks to @rraustad and @mjwitte for that.  @yujiex I'll do a quick build with latest develop as of right now to confirm things, but this should merge in this morning if it's all happy.

Myoldmopar commented:
Yep, all happy, merging this.  Thanks @yujiex and all

yujiex commented:
Thanks @Myoldmopar for compiling and merging! Also thanks @rraustad and @mjwitte for the comments and guidance.
